### ISSUE TITLE
Prepare hosting components for single-exe with integrated runtime

### ISSUE THREAD

**Issue by vitek-karas** [dotnet/core-setup#8299]
If one day the single-exe bundle integrates the hosting components and the runtime into the `apphost` itself and won't extract them to disk, the current design of native hosting will break.

Right now native hosting scenarios which "Attach" to a process with .NET Core runtime already running rely on this mechanism:
* Find the `hostfxr` module in the process and get a handle to it
* Get an export from the module
* Call the export to ask `hostfxr` to load the new managed component into the already running runtime (and get a pointer to some managed function back typically).

If the hosting and the runtime are all compiled/linked into the host there will be no `hostfxr` module in the process. So the above algorithm will break.

In the worst case, existing native hosting code might decide to fallback to search on disk, find a functional `hostfxr` there, load it into the process, which will in turn load `hostpolicy` and eventually the CoreCLR runtime itself. This might lead to having two instanced of the runtime in one process - in a semi working state. This is not expected behavior from the user and it's not a supported scenario. The code should actively try to prevent it.

The minimum solution to this will be to have a way to detect the runtime without relying on a search for known modules/exports. If the runtime is detected and there's no `hostfxr` simply fail the native hosting request. This would block these scenarios with clear and predictable error. Meaning we can reenable it in the future with a proper solution.

For 3.1 we should implement this block, so that in 5 we get a chance to correctly and fully implement the scenario. (Implementing the scenario for real in 3.1 is too risky and too big of a change, especially since we don't know how exactly the single-exe will look like yet).

**Comment by jkotas** [comment]
I do not think we need to worry about sharing the runtime and framework bundled in the single-exe with other dynamically loaded components. I think about the single-exe as static linking in C/C++. If you statically link C/C++ runtime or other libraries into your .exe, there is no way to share them with other dynamically loaded components either. The other dynamically loaded components have to bring their own.

**Comment by vitek-karas** [comment]
@jkotas that would mean we would sign up to support (if we support the scenario) two .NET Core runtimes in one process. That feels a bit risky. Note that this proposal is to actually actively block the ability to load anything into the single-exe runtime from the outside (since we don't know how to do that yet anyway).
We may decide that we keep that block in place indefinitely.

**Comment by jkotas** [comment]
> sign up to support (if we support the scenario) two .NET Core runtimes in one process

We can say that we are not supporting this in the documentation, without actually blocking this. There are many things that we say are unsupported (e.g. private reflection - that is super easy to use too) without being actively blocked.
